stacker/providers/aws/default.py

```diff
@@ -448,17 +452,34 @@ def generate_stack_policy_args(stack_policy=None):
 
 
 class ProviderBuilder(object):
-    """Implements a ProviderBuilder for the AWS provider."""
+    """Implements a Memoized ProviderBuilder for the AWS provider."""
 
     def __init__(self, region=None, **kwargs):
         self.region = region
         self.kwargs = kwargs
+        self.providers = {}
+        self.lock = Lock()
 
     def build(self, region=None, profile=None):
-        if not region:
-            region = self.region
-        session = get_session(region=region, profile=profile)
-        return Provider(session, region=region, **self.kwargs)
+        """Get or create the provider for the given region and profile."""
+
+        with self.lock:
+            # memoization lookup key derived by taking md5 of region + profile.
+            key = md5("{}{}".format(region, profile)).hexdigest()
```

Comment:
I think you can just use the [hash built-in](https://docs.python.org/2/library/functions.html#hash) to generate a consistent hash for the dict key:

```python
>>> hash("{}{}".format(region, profile))
-2792180382875253710
```

Reply:
fixed, couldn't we also just create a string key, like `"{}-{}".format(region, profile)`?

Reply:
True, just a string key would be fine as well.